reconcile/ocm_subscription_labels/integration.py

```diff
@@ -0,0 +1,281 @@
+        dry_run: bool,
+        current_cluster_states: ClusterStates,
+        desired_cluster_states: ClusterStates,
+    ) -> None:
+        for cluster_name, desired_cluster_state in desired_cluster_states.items():
+            current_cluster_state = current_cluster_states[cluster_name]
+            if desired_cluster_state == current_cluster_state:
+                continue
+
+            for label_to_add in (
+                desired_cluster_state.labels.keys()
+                - current_cluster_state.labels.keys()
+            ):
+                logging.info(
+                    [
+                        "create_cluster_subscription_label",
+                        cluster_name,
+                        f"{label_to_add}={desired_cluster_state.labels[label_to_add]}",
+                    ]
+                )
+                if not dry_run:
+                    add_subscription_labels(
+                        ocm_api=desired_cluster_state.ocm_api,
+                        cluster=desired_cluster_state.cluster_details,
+                        labels={
+                            label_to_add: desired_cluster_state.labels[label_to_add]
+                        },
+                    )
+
+            for label_to_rm in (
+                current_cluster_state.labels.keys()
+                - desired_cluster_state.labels.keys()
+            ):
+                logging.info(
+                    [
+                        "delete_cluster_subscription_label",
+                        cluster_name,
+                        f"{label_to_rm}={current_cluster_state.labels[label_to_rm]}",
+                    ]
+                )
+                if not dry_run:
+                    delete_subscription_labels(
+                        ocm_api=desired_cluster_state.ocm_api,
+                        cluster=desired_cluster_state.cluster_details,
+                        labels=[label_to_rm],
+                    )
+
+            for label_to_cmp in (
+                desired_cluster_state.labels.keys()
+                & current_cluster_state.labels.keys()
+            ):
+                if (
+                    desired_cluster_state.labels[label_to_cmp]
+                    == current_cluster_state.labels[label_to_cmp]
+                ):
+                    continue
+                logging.info(
+                    [
+                        "update_cluster_subscription_label",
+                        cluster_name,
+                        f"{label_to_cmp}={desired_cluster_state.labels[label_to_cmp]}",
+                    ]
+                )
+                if not dry_run:
+                    update_subscription_labels(
+                        ocm_api=desired_cluster_state.ocm_api,
+                        cluster=desired_cluster_state.cluster_details,
+                        labels={
+                            label_to_cmp: desired_cluster_state.labels[label_to_cmp]
+                        },
+                    )
```

Comment:
I'm selling my new generic differ here:

```python
            diff_result = diff_mappings(
                current_cluster_state.labels, desired_cluster_state.labels
            )

            for label_to_add, label in diff_result.add.items():
                logging.info(
                    [
                        "create_cluster_subscription_label",
                        cluster_name,
                        f"{label_to_add}={label}",
                    ]
                )
                if not dry_run:
                    add_subscription_labels(
                        ocm_api=desired_cluster_state.ocm_api,
                        cluster=desired_cluster_state.cluster_details,
                        labels={label_to_add: label},
                    )
            for label_to_rm, label in diff_result.delete.items():
                logging.info(
                    [
                        "delete_cluster_subscription_label",
                        cluster_name,
                        f"{label_to_rm}={label}",
                    ]
                )
                if not dry_run:
                    delete_subscription_labels(
                        ocm_api=desired_cluster_state.ocm_api,
                        cluster=desired_cluster_state.cluster_details,
                        labels=[label_to_rm],
                    )
            for label_to_update, diff_pair in diff_result.change.items():
                label = diff_pair.desired
                logging.info(
                    [
                        "update_cluster_subscription_label",
                        cluster_name,
                        f"{label_to_update}={label}",
                    ]
                )
                if not dry_run:
                    update_subscription_labels(
                        ocm_api=desired_cluster_state.ocm_api,
                        cluster=desired_cluster_state.cluster_details,
                        labels={label_to_update: label},
                    )
```

Reply:
I totally forget about the generic differ! Thanks for reminding me 🙇 
Added in [f7fbc31](https://github.com/app-sre/qontract-reconcile/pull/3714/commits/f7fbc317354060aa49851e5a725e9ff69d0cdbe4)